File: resilient/resilient-annotation-processor/src/main/java/ru/tinkoff/kora/resilient/annotation/processor/aop/TimeoutKoraAspect.java
package ru.tinkoff.kora.resilient.annotation.processor.aop;

import com.palantir.javapoet.AnnotationSpec;
import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.MethodUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.aop.annotation.processor.KoraAspect;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static com.palantir.javapoet.CodeBlock.joining;

public class TimeoutKoraAspect implements KoraAspect {

    private static final ClassName ANNOTATION_TYPE = ClassName.get("ru.tinkoff.kora.resilient.timeout.annotation", "Timeout");
    private static final ClassName EXHAUSTED_EXCEPTION = ClassName.get("ru.tinkoff.kora.resilient.timeout", "TimeoutExhaustedException");

    private final ProcessingEnvironment env;

    public TimeoutKoraAspect(ProcessingEnvironment env) {
        this.env = env;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Set.of(ANNOTATION_TYPE.canonicalName());
    }

    @Override
    public Set<ClassName> getSupportedAnnotationClassNames() {
        return Set.of(ANNOTATION_TYPE);
    }

    @Override
    public ApplyResult apply(ExecutableElement method, String superCall, AspectContext aspectContext) {
        if (MethodUtils.isPublisher(method)) {
            throw new ProcessingErrorException("Publisher methods are not supported", method);
        }
        final Optional<? extends AnnotationMirror> mirror = method.getAnnotationMirrors().stream().filter(a -> a.getAnnotationType().toString().equals(ANNOTATION_TYPE.canonicalName())).findFirst();
        final String timeoutName = mirror.flatMap(a -> a.getElementValues().entrySet().stream()
                .filter(e -> e.getKey().getSimpleName().contentEquals("value"))
                .map(e -> String.valueOf(e.getValue().getValue())).findFirst())
            .orElseThrow();

        var managerType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.timeout.TimeoutManager"));
        var fieldManager = aspectContext.fieldFactory().constructorParam(managerType, List.of());
        var metricsType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.timeout.TimeoutMetrics"));
        var fieldMetrics = aspectContext.fieldFactory().constructorParam(metricsType, List.of(AnnotationSpec.builder(Nullable.class).build()));
        var timeouterType = env.getTypeUtils().getDeclaredType(env.getElementUtils().getTypeElement("ru.tinkoff.kora.resilient.timeout.Timeout"));
        var fieldTimeout = aspectContext.fieldFactory().constructorInitialized(timeouterType,
            CodeBlock.of("$L.get($S)", fieldManager, timeoutName));

        final CodeBlock body;
        if (MethodUtils.isFuture(method)) {
            body = buildBodyFuture(method, superCall, timeoutName, fieldTimeout, fieldMetrics);
        } else {
            body = buildBodySync(method, superCall, fieldTimeout);
        }

        return new ApplyResult.MethodBody(body);
    }

    private CodeBlock buildBodySync(ExecutableElement method, String superCall, String timeoutName) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        if (MethodUtils.isVoid(method)) {
            return CodeBlock.builder().add("""
                $L.execute(() -> {
                    $L;
                    return null;
                });
                """, timeoutName, superMethod.toString()).build();
        } else {
            return CodeBlock.builder().add("""
                return $L.execute(() -> $L);
                """, timeoutName, superMethod.toString()).build();
        }
    }

    private CodeBlock buildBodyFuture(ExecutableElement method, String superCall, String timeoutName, String fieldTimeout, String fieldMetrics) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
                return $L.toCompletableFuture()
                    .orTimeout($L.timeout().toMillis(), $T.MILLISECONDS)
                    .exceptionallyCompose(_e -> {
                      var _cause = _e;
                      if (_cause instanceof $T ce) {
                          _cause = ce.getCause();
                      }
                      if(_cause instanceof $T) {
                        if($L != null) {
                            $L.recordTimeout($S, $L.timeout().toNanos());
                        }
                        return $T.failedFuture(new $T($S, "Timeout exceeded " + $L.timeout()));
                      } else {
                        return $T.failedFuture(_cause);
                      }
                    });""", superMethod.toString(), fieldTimeout, TimeUnit.class, CompletionException.class, TimeoutException.class,
            fieldMetrics, fieldMetrics, timeoutName, fieldTimeout, CompletableFuture.class,
            EXHAUSTED_EXCEPTION, timeoutName, fieldTimeout, CompletableFuture.class).build();
    }

    private CodeBlock buildBodyMono(ExecutableElement method, String superCall, String timeoutName, String fieldTimeout, String fieldMetrics) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
                return $L
                    .timeout($L.timeout())
                    .onErrorMap(e -> e instanceof $T, e -> new $T($S, "Timeout exceeded " + $L.timeout()))
                    .doOnError(e -> {
                        if(e instanceof $T && $L != null) {
                            $L.recordTimeout($S, $L.timeout().toNanos());
                        }
                    });
                """, superMethod.toString(), fieldTimeout, TimeoutException.class, EXHAUSTED_EXCEPTION, timeoutName,
            fieldTimeout, EXHAUSTED_EXCEPTION, fieldMetrics, fieldMetrics, timeoutName, fieldTimeout).build();
    }

    private CodeBlock buildBodyFlux(ExecutableElement method, String superCall, String timeoutName, String fieldTimeout, String fieldMetrics) {
        final CodeBlock superMethod = buildMethodCall(method, superCall);

        return CodeBlock.builder().add("""
                return $L
                    .timeout($L.timeout())
                    .onErrorMap(e -> e instanceof $T, e -> new $T($S, "Timeout exceeded " + $L.timeout()))
                    .doOnError(e -> {
                        if(e instanceof $T && $L != null) {
                            $L.recordTimeout($S, $L.timeout().toNanos());
                        }
                    });
                """, superMethod.toString(), fieldTimeout, TimeoutException.class, EXHAUSTED_EXCEPTION, timeoutName,
            fieldTimeout, EXHAUSTED_EXCEPTION, fieldMetrics, fieldMetrics, timeoutName, fieldTimeout).build();
    }

    private CodeBlock buildMethodCall(ExecutableElement method, String call) {
        return method.getParameters().stream().map(p -> CodeBlock.of("$L", p)).collect(joining(", ", call + "(", ")"));
    }
}

File: resilient/resilient-symbol-processor/src/main/kotlin/ru/tinkoff/kora/resilient/symbol/processor/aop/TimeoutKoraAspect.kt
package ru.tinkoff.kora.resilient.symbol.processor.aop

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.MemberName
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlow
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlux
import ru.tinkoff.kora.ksp.common.FunctionUtils.isCompletionStage
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFuture
import ru.tinkoff.kora.ksp.common.FunctionUtils.isMono
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.FunctionUtils.isVoid
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.concurrent.CompletionStage
import java.util.concurrent.Future

@KspExperimental
class TimeoutKoraAspect(val resolver: Resolver) : KoraAspect {

    companion object {
        private val ANNOTATION_TYPE = ClassName("ru.tinkoff.kora.resilient.timeout.annotation", "Timeout")
        val MEMBER_CALLABLE = MemberName("java.util.concurrent", "Callable")
        val timeoutMember = MemberName("kotlinx.coroutines", "withTimeout")
        val timeoutCancelMember = MemberName("kotlinx.coroutines", "TimeoutCancellationException")
        val flowMember = MemberName("kotlinx.coroutines.flow", "flow")
        val emitMember = MemberName("kotlinx.coroutines.flow", "emitAll")
        val startMember = MemberName("kotlinx.coroutines.flow", "onStart")
        val whileMember = MemberName("kotlinx.coroutines.flow", "takeWhile")
        val systemMember = MemberName("java.lang", "System")
        val atomicMember = MemberName("java.util.concurrent.atomic", "AtomicLong")
        val timeoutKoraMember = MemberName("ru.tinkoff.kora.resilient.timeout", "TimeoutExhaustedException")
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(ANNOTATION_TYPE.canonicalName)
    }

    override fun apply(ksFunction: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
        if (ksFunction.isFuture()) {
            throw ProcessingErrorException("@Timeout can't be applied for types assignable from ${Future::class.java}", ksFunction)
        } else if (ksFunction.isCompletionStage()) {
            throw ProcessingErrorException("@Timeout can't be applied for types assignable from ${CompletionStage::class.java}", ksFunction)
        } else if (ksFunction.isMono()) {
            throw ProcessingErrorException("@Timeout can't be applied for types assignable from ${CommonClassNames.mono}", ksFunction)
        } else if (ksFunction.isFlux()) {
            throw ProcessingErrorException("@Timeout can't be applied for types assignable from ${CommonClassNames.flux}", ksFunction)
        }

        val annotation = ksFunction.findAnnotation(ANNOTATION_TYPE)!!

        val timeoutName = annotation.findValue<String>("value")!!

        val metricType = resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.timeout.TimeoutMetrics")!!.asType(listOf()).makeNullable()
        val fieldMetric = aspectContext.fieldFactory.constructorParam(metricType, listOf())
        val managerType = resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.timeout.TimeoutManager")!!.asType(listOf())
        val fieldManager = aspectContext.fieldFactory.constructorParam(managerType, listOf())
        val fieldTimeout = aspectContext.fieldFactory.constructorInitialized(
            resolver.getClassDeclarationByName("ru.tinkoff.kora.resilient.timeout.Timeout")!!.asType(listOf()),
            CodeBlock.of("%L[%S]", fieldManager, timeoutName)
        )

        val body = if (ksFunction.isFlow()) {
            buildBodyFlow(ksFunction, superCall, timeoutName, fieldTimeout, fieldMetric)
        } else if (ksFunction.isSuspend()) {
            buildBodySuspend(ksFunction, superCall, timeoutName, fieldTimeout, fieldMetric)
        } else {
            buildBodySync(ksFunction, superCall, fieldTimeout)
        }

        return KoraAspect.ApplyResult.MethodBody(body)
    }

    private fun buildBodySync(
        method: KSFunctionDeclaration, superCall: String, timeoutName: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        return if (method.isVoid()) {
            CodeBlock.builder().add(
                """
                    %L.execute( %M { %L })
                    """.trimIndent(), timeoutName, MEMBER_CALLABLE, superMethod.toString()
            ).build()
        } else {
            CodeBlock.builder().add(
                """
                    return %L.execute( %M { %L })
                    """.trimIndent(), timeoutName, MEMBER_CALLABLE, superMethod.toString()
            ).build()
        }
    }

    private fun buildBodySuspend(
        method: KSFunctionDeclaration, superCall: String, timeoutName: String, fieldTimeout: String, fieldMetric: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            try {
                  return %M(%L.timeout().toMillis()) {
                      %L
                  }
            } catch (e: %M) {
                %L?.recordTimeout(%S, %L.timeout().toNanos())
                throw %M(%S, "Timeout exceeded " + %L.timeout())
            }
          """.trimIndent(), timeoutMember, fieldTimeout, superMethod.toString(), timeoutCancelMember,
            fieldMetric, timeoutName, fieldTimeout, timeoutKoraMember, timeoutName, fieldTimeout
        ).build()
    }

    private fun buildBodyFlow(
        method: KSFunctionDeclaration, superCall: String, timeoutName: String, fieldTimeout: String, fieldMetric: String
    ): CodeBlock {
        val superMethod = buildMethodCall(method, superCall)
        return CodeBlock.builder().add(
            """
            val limit = %M()
            return %M { %M(%L) }
                .%M { limit.set(%M.nanoTime() + %L.timeout().toNanos()) }
                .%M {
                    val current = %M.nanoTime()
                    if (current > limit.get()) {
                        %L?.recordTimeout(%S, %L.timeout().toNanos())
                        throw %M(%S, "Timeout exceeded " + %L.timeout())
                    } else {
                        false
                    }
                }
            """.trimIndent(),
            atomicMember, flowMember, emitMember, superMethod.toString(), startMember, systemMember,
            fieldTimeout, whileMember, systemMember, fieldMetric, timeoutName, fieldTimeout, timeoutKoraMember, timeoutName, fieldTimeout,
        ).build()
    }

    private fun buildMethodCall(method: KSFunctionDeclaration, call: String): CodeBlock {
        return CodeBlock.of(method.parameters.asSequence().map { p -> CodeBlock.of("%L", p) }.joinToString(", ", "$call(", ")"))
    }
}
